### PR TITLE
Option for excluding rename commits from last update

### PR DESCRIPTION
**Pull request by skywarth**
Discussed and proposed in #147 

Did a couple of tests using one of my repositories where the aforementioned problem is present. Looking good so far.

This is just a draft PR, do not merge yet. Populate the TODO below if I'm missing something

TODO:

- [X] Manual test on an actual repository, via `mkdocs serve`
- [ ] Automation tests defined in fixture
  - [ ] Maybe define a new fixture for this case
- [x] Update the documentation to include description for this new option

### CODE REVIEW

**Comment by timvink** [comment]
I don't think we need an option for this. If there any use case where you _would_ want this?

**Comment by skywarth** [comment]
> I don't think we need an option for this. If there any use case where you _would_ want this?

You mean this should be the default behavior, without an option? 

I think the feature itself would be useful, because in knowledge-base and documentation projects there are often quite a lot of folder hierarchy changes, and if we consider these changes as 'last update' it could be misleading. Say for an API documentation; it was written in 2020 and never updated, which is deprecated today. And lately it was decided to be moved to another folder, and now it shows as it is updated recently, and people may think it is up-to-date, whereas it is not.

The reason to have it as option and default behavior could be; it brings an additional overhead since it runs with `--follow` or `follow=True` and some people might not want such overhead which might cripple the performance a bit where it is imperative.

**Comment by timvink** [comment]
OK I had to dig back into the issue. 

So if we use `--follow`, there is this really nasty back that the date will be incorrect if there are any empty files committed in the git history (https://blog.plover.com/prog/git-log-follow.html). Hence I _was_ planning to make that setting optional, so users basically have to choose between incorrect creation/revision dates if there are file names, vs incorrect creation/revision dates if there are any empty files committed.

I _do_ want to have sensible defaults. Here's my proposal:

- We add a new plugin option `follow`, that defaults to `True`. This is a breaking change, because revision dates might changes (follow was already enabled for creation dates). 
- We add the diff filter `r` for revision dates. Does that diff filter need `follow` to be True, or can we just leave it enabled always? If possible, we can remove the conditional on follow like you have it now.
- While we're making breaking changes, let's also add the `--ignore-all-space` and `--ignore-blank-lines` to the revision git log command. I think it's sensible, as those are not _real_ edits. I can document it in the release notes.

Would you be willing to make the changes? I can take care of the documentation

File: src/mkdocs_git_revision_date_localized_plugin/util.py [hunk 1]
```diff
@@ -92,8 +92,15 @@ def get_git_commit_timestamp(
...
+                if self.config.get('last_update_exclude_renames'):
+                    # We can remove if-else statement and do a one-liner as well
+                    diff_filter_param = "r"
+                    follow_param = True
+                else:
+                    diff_filter_param = ""
+                    follow_param = False
                 commit_timestamp = git.log(
-                    realpath, date="unix", format="%at", n=1, no_show_signature=True
+                    realpath, date="unix", format="%at", diff_filter=diff_filter_param, n=1, no_show_signature=True, follow=follow_param
```
Comment (timvink) [inline src/mkdocs_git_revision_date_localized_plugin/util.py:103]: Note that with follow enabled, you will get multiple commits, and you need to take the last one. See the logic used for the creation date.

**Comment by skywarth** [comment]
Sure @timvink, though I didn't fully wrap my head around it I'll try to implement these. If it gets way above my head I'll let you know :smile:

File: src/mkdocs_git_revision_date_localized_plugin/util.py (hunk 1 above)
Comment (skywarth) [reply src/mkdocs_git_revision_date_localized_plugin/util.py]: Doesn't `n=1` also achieve the same desired effect as selecting the last one using `commit_timestamp = commit_timestamp.split()[-1]` ?

I saw that and though maybe it is just awaiting refactoring.

If it doesn't achieve the same result, I can do it like the creation date.

**Comment by skywarth** [comment]
Hi @timvink, just pushed the changes:

> * [X] We add a new plugin option `follow`, that defaults to `True`. This is a breaking change, because revision dates might changes (follow was already enabled for creation dates).
> * [X] We add the diff filter `r` for revision dates. No plugin option, I think ignoring renames makes sense for both revision and creation dates. 
> * [X] While we're making breaking changes, let's also add the `--ignore-all-space` and `--ignore-blank-lines` to the revision git log command. I think it's sensible, as those are not _real_ edits. I can document it in the release notes.

About the `"r"` and `--follow`
> "Does that diff filter = 'r'' need --follow? or can we just always add the diff filter? Maybe we can remove the conditional on follow like you have it now."

We can always add the `diff-filter="r"`, when it is used along with `--follow` it essentially takes moving operations into account too. When it is used without the `--follow` it only tracks actual renames; renames that are limited to current directory, moving operations are not included since the file is not **tracked/followed**.

I got one question though; why did we add diff-filter "r" for creation dates as well? I was trying to come up with a case where it would make a difference but couldn't think of any, could you describe a case for it please?

And for the manual tests, here's the difference between `master` branch and this branch's outputs, based on one of my repositories:
- [problems.md](https://github.com/skywarth/annals-of-disco/blob/master/docs/information-technology/problems.md)
    - `master` branch, [(deployed)](https://skywarth.github.io/annals-of-disco/information-technology/problems/): Revision date  September 24, 2024. Creation date July 16, 2022.
    - `PR` branch, `version_history_follow_enabled: true`: Revision date July 16, 2022. Creation date July 16, 2022.
    - `PR` branch, `version_history_follow_enabled: false`: Revision date September 24, 2024. Creation date September 24, 2024.
- [great-advertisements.md](https://github.com/skywarth/annals-of-disco/blob/master/docs/marketing/great-advertisements.md)
    - `master` branch, [(deployed)](https://skywarth.github.io/annals-of-disco/information-technology/problems/): Revision date September 28, 2024. Creation date August 10, 2024.
    - `PR` branch, `version_history_follow_enabled: true`: September 28, 2024. Creation date August 10, 2024.
    - `PR` branch, `version_history_follow_enabled: false`: September 28, 2024. Creation date September 24, 2024.

Please note September 24, 2024 is the date I moved all files to another directory. 

In conclusion: based on the test outputs, I think it achieves what we are looking for.

Let me know if you need additional changes or tests. I'm here to help. 

Thank you

File: src/mkdocs_git_revision_date_localized_plugin/util.py [hunk 2]
```diff
@@ -93,8 +95,11 @@ def get_git_commit_timestamp(
             else:
                 # Latest commit touching a specific file
                 commit_timestamp = git.log(
-                    realpath, date="unix", format="%at", n=1, no_show_signature=True
+                    realpath, date="unix", format="%at",
+                    diff_filter="r", n=1, no_show_signature=True, follow=follow_option,
```
[the remark this answers is not in the document]
Comment (timvink) [reply src/mkdocs_git_revision_date_localized_plugin/util.py:99]: For speed I assume? 

image

**Comment by timvink** [comment]
Nice work!

> I got one question though; why did we add diff-filter "r" for creation dates as well? I was trying to come up with a case where it would make a difference but couldn't think of any, could you describe a case for it please? 

I assume speed. Not sure anymore tbh. 

> Let me know if you need additional changes or tests. I'm here to help.

Cool!

- can you rename `version_history_follow_enabled` to `enable_git_follow` 
- if you want, you can already update `docs/options.md`

**Comment by skywarth** [comment]
Certainly @timvink, I'm slightly busy nowadays but no worries I'll let you know once I update the PR, stay tuned. Thank you.

File: src/mkdocs_git_revision_date_localized_plugin/util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (skywarth) [reply src/mkdocs_git_revision_date_localized_plugin/util.py:99]: Yeah most probably. It was already there, I didn't add it.

**Comment by skywarth** [comment]
Hi @timvink. Sorry for the delay, I was busy with relocating. Can you take a quick look and see if it's up to par?

**Comment by timvink** [comment]
Great stuff, thanks for all your effort to push this!

**Comment by timvink** [comment]
Will make a release this weekend

**Comment by timvink** [comment]
Released! https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/releases/tag/v1.3.0

**Comment by skywarth** [comment]
> Released! https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/releases/tag/v1.3.0

Brilliant, cheers!
